src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,154 @@
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(value);
+            if (head == null) {
+                head = newNode;
+                tail = newNode;
+            } else {
+                newNode.next = head;
+                head.prev = newNode;
+                head = newNode;
+            }
+            size++;
+        } else if (index == size) {
+            add(value);
+        } else {
+            Node<T> newNode = new Node<>(value);
+            Node<T> currentNode = head;
+            for (int i = 0; i < index - 1; i++) {
+                currentNode = currentNode.next;
+            }
+            newNode.next = currentNode.next;
+            newNode.prev = currentNode;
+            currentNode.next.prev = newNode;
+            currentNode.next = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T elementOfList : list) {
+            add(elementOfList);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("invalid index " + index);
+        }
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkingIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkingIndex(index);
+        Node<T> currentNode = head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T oldValue = currentNode.value;
+        unlink(currentNode);
+        size--;
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == null ? object == null : currentNode.value.equals(object)) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkingIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("invalid index " + index);
+        }
+    }
+
+    private void unlink(Node<T> node) {
```

Comment:
The 'unlink' method should handle the case where the 'node' is either the head or the tail of the list. Make sure to update 'head' and 'tail' references accordingly.